### PR TITLE
Added Feature to deploy previews of pull requests

### PR DESCRIPTION
**Pull request by devanshkansagra**
# Pull Request Preview Deployment to GitHub Pages
## Acceptance Criteria fulfillment

- [x] Deploy every pull request builds of monorepo to a unique url
- [x] Remove Deployment when PR is merged and closed

## Video/Screenshots
image

### CODE REVIEW

File: .github/workflows/cd-pr.yml
Comment (Spiral-Memory) [inline .github/workflows/cd-pr.yml:9]: Once check if can remove any permission from here (as this workflow will have write permission

File: .github/workflows/ci-pr.yml
Comment (Spiral-Memory) [inline .github/workflows/ci-pr.yml:7]: Same (but here it is less risky)

File: .github/workflows/ci-pr.yml
Comment (Spiral-Memory) [inline .github/workflows/ci-pr.yml:12]: Could you please explain this?

File: .github/workflows/ci-pr.yml
Comment (Spiral-Memory) [inline .github/workflows/ci-pr.yml:18]: Use DOCS_BASE_URL for consistency (also make sure the application is expecting the same env variable

File: .github/workflows/ci-pr.yml
Comment (Spiral-Memory) [inline .github/workflows/ci-pr.yml:24]: Remove outputs as discussed and tested (as it has no use)

File: .github/workflows/pr-cleanup.yml
Comment (Spiral-Memory) [inline .github/workflows/pr-cleanup.yml:16]: Make sure it must not run if the PR is closed / merged before it got apporved

File: .github/workflows/cd-pr.yml
Comment (Spiral-Memory) [inline .github/workflows/cd-pr.yml:1]: Let's rename cd-pr.yml as deploy-pr-preview.yml and ci as build-pr-preview.yml

CI/CD name is confusing as other workflow is also a part of CI

File: .github/workflows/ci-pr.yml
Comment (devanshkansagra) [reply .github/workflows/ci-pr.yml:12]: Actually I forgot about this, I also didn't know now why I used it, but during testing it was not creating any issues

File: .github/workflows/ci-pr.yml
Comment (devanshkansagra) [reply .github/workflows/ci-pr.yml:24]: done

File: .github/workflows/ci-pr.yml
Comment (Spiral-Memory) [reply .github/workflows/ci-pr.yml:12]: It ensures that only a specified number of workflow runs are executed at the same time, and it can help prevent redundant or conflicting runs, saving both time and resources.

Looks fine !

File: .github/workflows/cd-pr.yml
Comment (devanshkansagra) [reply .github/workflows/cd-pr.yml:9]: we will need these permissions for deploying builds to github-pages. permissions:content: write

Comment (Spiral-Memory) [reply .github/workflows/cd-pr.yml:9]: pages: write
id-token: write

Are these necessary ?

**Review by Spiral-Memory — APPROVED** [review]
Approving to test build !

File: .github/workflows/cd-pr.yml
Comment (devanshkansagra) [reply .github/workflows/cd-pr.yml:9]: pages write gives permissions to deploy from github pages and id-token I don't know, it was there in build-lint so I used it

Comment (Spiral-Memory) [reply .github/workflows/cd-pr.yml:9]: Try removing permissions one by one and test if it is still working fine, let's allow as less permissions as we can

Comment (devanshkansagra) [reply .github/workflows/cd-pr.yml:9]: done

File: .github/workflows/ci-pr.yml
Comment (devanshkansagra) [reply .github/workflows/ci-pr.yml:7]: removed and working well

File: .github/workflows/ci-pr.yml
Comment (devanshkansagra) [reply .github/workflows/ci-pr.yml:18]: changed

File: .github/workflows/pr-cleanup.yml
Comment (devanshkansagra) [reply .github/workflows/pr-cleanup.yml:16]: fixed

File: .github/workflows/cd-pr.yml
Comment (devanshkansagra) [reply .github/workflows/cd-pr.yml:1]: changed the name

**Review by Spiral-Memory — APPROVED** [review]

**Comment by Spiral-Memory** [comment]
Thank you so much for the contribution, @devanshkansagra.  
I truly appreciate the time and effort you put into this PR, as it was a much-needed feature to allow maintainers to test changes visually before merging. Since I worked with you on this, I know how much time, effort, and research went into building and testing it on a dummy repo.

With all good intentions, I am merging it to see if everything works as expected.

**Note:** We understand the security risks associated with the `pull_request_target` and `workflow_run` triggers, and we have done our best to maintain a balance between convenience and security. However, if any security incidents arise due to this, we are ready to revert the changes at any time.

Thanks again, @devanshkansagra. Happy contributing!

**Pull request merged**
